server/v2/cometbft/handlers/defaults.go

```diff
@@ -0,0 +1,160 @@
+package handlers
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	consensusv1 "cosmossdk.io/api/cosmos/consensus/v1"
+	abci "cosmossdk.io/api/tendermint/abci"
+	corecontext "cosmossdk.io/core/context"
+	"cosmossdk.io/core/transaction"
+	"cosmossdk.io/server/v2/cometbft/mempool"
+	"cosmossdk.io/server/v2/core/appmanager"
+	"google.golang.org/protobuf/proto"
+)
+
+// PrepareHandler passes in the list of Txs that are being proposed. The app can then do stateful operations
+// over the list of proposed transactions. It can return a modified list of txs to include in the proposal.
+type PrepareHandler[T transaction.Tx] func(context.Context, AppManager[T], []T, proto.Message) ([]T, error)
+
+// ProcessHandler is a function that takes a list of transactions and returns a boolean and an error.
+// If the verification of a transaction fails, the boolean is false and the error is non-nil.
+type ProcessHandler[T transaction.Tx] func(context.Context, AppManager[T], []T, proto.Message) error
+
+type AppManager[T transaction.Tx] interface {
+	ValidateTx(ctx context.Context, tx T, execMode corecontext.ExecMode) (appmanager.TxResult, error)
+	Query(ctx context.Context, version uint64, request appmanager.Type) (response appmanager.Type, err error)
+}
+
+type DefaultProposalHandler[T transaction.Tx] struct {
+	mempool    mempool.Mempool[T]
+	txSelector TxSelector[T]
+}
+
+func NewDefaultProposalHandler[T transaction.Tx](mp mempool.Mempool[T]) *DefaultProposalHandler[T] {
+	return &DefaultProposalHandler[T]{
+		mempool:    mp,
+		txSelector: NewDefaultTxSelector[T](),
+	}
+}
+
+func (h *DefaultProposalHandler[T]) PrepareHandler() PrepareHandler[T] {
+	return func(ctx context.Context, app AppManager[T], txs []T, req proto.Message) ([]T, error) {
+
+		abciReq, ok := req.(*abci.RequestPrepareProposal)
+		if !ok {
+			return nil, fmt.Errorf("invalid request type: %T", req)
+		}
+
+		var maxBlockGas uint64
+
+		res, err := app.Query(ctx, 0, &consensusv1.QueryParamsRequest{})
+		if err != nil {
+			return nil, err
+		}
+
+		paramsResp, ok := res.(*consensusv1.QueryParamsResponse)
+		if !ok {
+			return nil, fmt.Errorf("failed to query consensus params")
+		}
+
+		if b := paramsResp.GetParams().Block; b != nil {
+			maxBlockGas = uint64(b.MaxGas)
+		}
+
+		defer h.txSelector.Clear()
+
+		// TODO: can we assume nil mempool is NoOp?
+		// If the mempool is nil or NoOp we simply return the transactions
+		// requested from CometBFT, which, by default, should be in FIFO order.
+		//
+		// Note, we still need to ensure the transactions returned respect req.MaxTxBytes.
+		if h.mempool == nil {
```

Comment:
We cannot assume nil == NoOp, in fact NoOp is an explicit mempool type. Maybe we can check it's type. But that would require a mempool import.

Reply:
Currently we return the NoOp handler when `if h.mempool == nil || isNoOp`. What would be the difference between a nil mempool and the no op mempool?

Reply:
since we cant disable the mempool from comets side the noop return txs. We would need to check nil mempool in processproposal in abci to make it work in the way you are expecting